Python/Module5_OddsAndEnds/Writing_Good_Code.ipynb

```diff
@@ -0,0 +1,1028 @@
+    "\n",
+    "It is is important to review the [abstract base classes (abc's) for Python's collections](https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes) (yes, I am encouraging you to learn your abc's). This will help you classify categories of types generally, based on their [special methods](http://www.pythonlikeyoumeanit.com/Module4_OOP/Special_Methods.html). Two of the most common abc's are `Iterable`: any class that supports the iteration protocol, and `Sequence`: any collection that has a length (via `__len__`) and supports the get-item syntax (via `__getitem__`)."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<div class=\"alert alert-info\">\n",
+    "\n",
+    "**Reading Comprehension: Type Hinting**\n",
+    "\n",
+    "Read through the following function and annotate its signature with type-hints. Try to make your type-hints sufficiently general here.\n",
+    "\n",
+    "```python\n",
+    "def get_first_and_last(x):\n",
+    "    \"\"\" Returns the first and last elements of `x`. `x` is \n",
+    "    assumed to be non-empty\n",
+    "    \"\"\" \n",
+    "    return (x[0], x[-1])\n",
+    "```\n",
+    "\n",
+    "Here are some examples of this function in action. Be sure that your type-hint captures all of this diversity.\n",
+    "\n",
+    "```python\n",
+    ">>> get_first_and_last(\"hello\")\n",
+    "('h', 'o')\n",
+    "\n",
+    ">>> get_first_and_last([0, 1, 2, 3])\n",
+    "(0, 3)\n",
+    "\n",
+    ">>> get_first_and_last((True, False))\n",
+    "(True, False)\n",
+    "```\n",
+    "\n",
+    "</div>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<div class=\"alert alert-info\">\n",
+    "\n",
+    "**Takeaway:**\n",
+    "\n",
+    "It is un-Pythonic to be overly strict when type hinting. Python handles typing flexibly - it uses [duck typing](https://en.wikipedia.org/wiki/Duck_typing) - and our type hints should follow suite. Rely on abstract base classes when possible to provide the appropriate guidelines for a particular variable type. \n",
+    "\n",
+    "</div>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Documentation Styles\n",
+    "\n",
+    "[Davis King](https://twitter.com/nulhom?lang=en), a prolific and talented open source developer, is the creator of the [Dlib C++ library](http://dlib.net/). Among the library's major features, like machine learning algorithms, Davis lists its *documentation* as the first and foremost feature. In fact he says about Dlib:\n",
```

Comment:
Stylisticlally, dlib (lowercase) is preferred unless it's starting a sentence.

....In fact**,** he says <-- comma